ppcs/ppcTODO-signature-named-parameters.md

```diff
@@ -0,0 +1,153 @@
+
+One example comes from [`IPC::MicroSocket::Server`](https://metacpan.org/pod/IPC::MicroSocket::Server). This is currently using the `Sublike::Extended` module (see "Prototype Implementation" below) to provide the syntax. The actual module also uses `Object::Pad` to provide the `method` keyword; this example is paraphrased to avoid that. This module requires a `path` named argument, and optionally takes a `listen` argument, defaulting its value to 5 if the caller did not provide a defined value.
+
+```perl
+extended sub new_unix ( $class, :$path, :$listen //= 5 )
+{
+    ...
+}
+```
+
+This replaces the previous version of the code which handled arguments by the more traditional approach of assigning into a hash:
+
+```perl
+sub new_unix ( $class, %args )
+{
+    my $path   = $args{path};
+    my $listen = $args{listen} // 5;
+    ...
+}
+```
+
+Already the new code is shorter and more consise. Additionally it contains error checking that complains about missing mandatory keys, or unrecognised keys, which the previous version of the code did not include.
+
+Another example, this time from [`Text::Treesitter::QueryCursor`](https://metacpan.org/pod/Text::Treesitter::QueryCursor). This method takes an optional argument named `multi`, tested for truth. If absent it should default to false.
+
+```perl
+sub next_match_captures ( $self, :$multi = 0 )
+{
+    ...
+}
+```
+
+The previous version of this code did include complaints about unrecognised keys, and was rather longer because of it:
+
+```perl
+sub next_match_captures ( $self, %options )
+{
+    my $multi = delete $options{multi};
+    keys %options and
+        croak "Unrecognised options to ->next_captures: " . join( ", ", keys %options );
+
+    ...
+}
+```
+
+## Prototype Implementation
+
+The [`XS::Parse::Sublike`](https://metacpan.org/pod/XS::Parse::Sublike) module contains parsing to allow third-party syntax modules to parse subroutine-like constructions, and includes a parser for named parameters already having this syntax. These are also made available to regular subroutines via the `extended` keyword provided by [`Sublike::Extended`](https://metacpan.org/pod/Sublike::Extended).
+
+## Future Scope
+
+* If the Metaprogramming API (PPC0022) gains introspection abilities to enquire about subroutine signature parameters, further consideration will need to be made in that API on how to represent the extra kinds of parameters added by this specification.
+
+## Rejected Ideas
+
+* This specification intentionally makes no changes to the call-site of any subroutines using named parameters.
+
+## Open Issues
+
+* How to specify a named parameter whose name is anything other than the name of the lexical variable into which its value is assigned? This is more of note when considering that traditional argument handling techniques involving assignment into hashes can handle more styles of name that would be invalid for lexical variables - for example, names including hyphens.
```

Comment:
Another thing to consider here is multiple names for the same parameter, eg, your make_colour() might accept `red`, `green` and `blue` or `r`, `g` and `b`.

Reply:
> Another thing to consider here is multiple names for the same parameter, eg, your make_colour() might accept `red`, `green` and `blue` or `r`, `g` and `b`.

I know that in PDL and R (e.g., `ggplot2`), it's typical to normalise named parameters to accept `colour` and `color`. Though I can see this kind of feature being handled via an attribute (or keep the custom code that is already there).

Raku's synopsis that I linked in a comment below gives the example of:

```raku
sub globalize (:g(:global($gl))) {...}
```

where the variable is `$gl`, but the named key is either `global` or `g`.

Reply:
Here's a possible approach:

```perl
sub f(
    # Internally as C<$colour> with external API of I<colour> and I<color>.
    :$colour :name(color),

    # Internally as C<$gl> with external API of I<global> and I<g>.
    # NOTE that C<$gl> does not have a colon (C<:>) preceding it.
    # The :name() indicates that it is not positional.
    $gl      :name(global g),

) {
    say "Color: $colour, Global: $gl";
}

sub c(
    :$red   :name(r), # both red   and r
    :$green :name(g), # both green and g
    :$blue  :name(b), # both blue  and b
) {
    ...
}
```

This aliasing approach does not have to be examined within this PPC and the
contents of this PPC don't conflict with it.

Reply:
@zmughal you are right this goes further then this PPC, there are tons of things ...

it depends on what we want to do with Perl itself. Focus on readability? Focus on catching train? Focus on defining strong side where other languages cannot compete yet? Focus in expressing implementation? Focus on expressing intention? 

Regarding readability: I suggest to read this: https://www.scientificamerican.com/blog/frontiers-for-young-minds/the-reading-brain/

there are also studies related to readability of code - though every such article/study will get opposition in style "I'm against this because I can read my code I just wrote"

So regarding this PPC:
- will it bring short-time improvement? yes
- will it bring long-time improvement? syntax-wise I'm skeptical, but I think it will bring enough improvements in core to make better syntax improvement possible

What this PPC is missing (according me) is comparison how other languages deal with problem and evaluation of pros/cons of such solution. At least first part I did some time ago: https://github.com/happy-barney/perl-wish-list/blob/master/named-arguments/spec/named-arguments-in-other-languages.md

Reply:
@zmughal 
> Here's a possible approach:
> 
> ```perl
> sub f(
>     # Internally as C<$colour> with external API of I<colour> and I<color>.
>     :$colour :name(color),

Yes, an attribute on the parameter would seem to be a sensible attack at that. We don't currently have an ability to put attributes in positions like that, but `XS::Parse::Sublike` permits that sort of thing. It's the subject of a different PPC of mine that I'm still in the middle of drafting. Perhaps I can extend the wording of this one to at least hint at it though.